File: opencart/algovoi/catalog/controller/payment/algovoi.php
<?php
/**
 * AlgoVoi: https://github.com/chopmob-cloud/AlgoVoi-Platform-Adapters
 * Licensed under the Business Source License 1.1 - see LICENSE for details.
 */
namespace Opencart\Catalog\Controller\Extension\Algovoi\Payment;

class Algovoi extends \Opencart\System\Engine\Controller {

    // All hosted networks with display labels
    private const ALL_NETWORKS = [
        'algorand_mainnet' => 'Algorand &mdash; USDC',
        'voi_mainnet'      => 'VOI &mdash; aUSDC',
        'hedera_mainnet'   => 'Hedera &mdash; USDC',
        'stellar_mainnet'  => 'Stellar &mdash; USDC',
        'base_mainnet'     => 'Base &mdash; USDC',
        'solana_mainnet'   => 'Solana &mdash; USDC',
        'tempo_mainnet'    => 'Tempo &mdash; USDC',
    ];

    private const NET_KEYS = [
        'algorand_mainnet' => 'algorand',
        'voi_mainnet'      => 'voi',
        'hedera_mainnet'   => 'hedera',
        'stellar_mainnet'  => 'stellar',
        'base_mainnet'     => 'base',
        'solana_mainnet'   => 'solana',
        'tempo_mainnet'    => 'tempo',
    ];

    public function index(): string {
        $this->load->language('extension/algovoi/payment/algovoi');
        $data['language'] = $this->config->get('config_language');

        // Build enabled_networks: default all enabled if config never saved
        $enabled = [];
        foreach (self::NET_KEYS as $net => $short) {
            $val = $this->config->get('payment_algovoi_net_' . $short);
            if ($val === null || $val === '' || $val === '1' || $val === 1) {
                $enabled[$net] = self::ALL_NETWORKS[$net];
            }
        }
        if (empty($enabled)) $enabled = self::ALL_NETWORKS;
        $data['enabled_networks'] = $enabled;

        return $this->load->view('extension/algovoi/payment/algovoi', $data);
    }

    public function confirm(): void {
        $this->load->language('extension/algovoi/payment/algovoi');
        $json = [];

        if (!isset($this->session->data['order_id'])) {
            $json['redirect'] = $this->url->link('checkout/failure', 'language=' . $this->config->get('config_language'), true);
            $this->response->addHeader('Content-Type: application/json');
            $this->response->setOutput(json_encode($json));
            return;
        }

        $this->load->model('checkout/order');
        $order = $this->model_checkout_order->getOrder($this->session->data['order_id']);

        if (!$order) {
            $json['redirect'] = $this->url->link('checkout/failure', 'language=' . $this->config->get('config_language'), true);
            $this->response->addHeader('Content-Type: application/json');
            $this->response->setOutput(json_encode($json));
            return;
        }

        $tenant_id = $this->config->get('payment_algovoi_tenant_id');
        $api_key   = $this->config->get('payment_algovoi_admin_api_key');
        // FIX: use trim() not db->escape() — this is API data, not SQL; whitelist is the real guard
        $network   = isset($_POST['algovoi_network']) ? trim($_POST['algovoi_network']) : ($this->config->get('payment_algovoi_preferred_network') ?: 'algorand_mainnet');
        $enabled_confirm = [];
        foreach (self::NET_KEYS as $net => $short) {
            $val = $this->config->get('payment_algovoi_net_' . $short);
            if ($val === null || $val === '' || $val === '1' || $val === 1) $enabled_confirm[] = $net;
        }
        if (empty($enabled_confirm)) $enabled_confirm = array_keys(self::ALL_NETWORKS);
        if (!in_array($network, $enabled_confirm, true)) $network = $enabled_confirm[0];

        $api_base = rtrim($this->config->get('payment_algovoi_api_base_url') ?: 'https://cloud.algovoi.co.uk', '/');

        $payload = json_encode([
            'amount'             => round((float)$order['total'], 2),
            'currency'           => strtoupper($order['currency_code']),
            'label'              => 'Order #' . $order['order_id'],
            'preferred_network'  => $network,
            'redirect_url'       => $this->url->link('extension/algovoi/payment/algovoi.callback', 'language=' . $this->config->get('config_language'), true),
            'expires_in_seconds' => 3600,
        ]);

        $ch = curl_init($api_base . '/v1/payment-links');
        curl_setopt_array($ch, [
            CURLOPT_POST            => true,
            CURLOPT_POSTFIELDS      => $payload,
            CURLOPT_RETURNTRANSFER  => true,
            CURLOPT_TIMEOUT         => 30,
            CURLOPT_SSL_VERIFYPEER  => true,   // FIX: explicit SSL verification
            CURLOPT_SSL_VERIFYHOST  => 2,
            CURLOPT_HTTPHEADER      => [
                'Content-Type: application/json',
                'Authorization: Bearer ' . $api_key,
                'X-Tenant-Id: ' . $tenant_id,
            ],
        ]);
        $response  = curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        $body = json_decode($response, true);

        if (($http_code === 200 || $http_code === 201) && !empty($body['checkout_url'])) {
            // SSRF guard: checkout URL must be on the same host as api_base
            $checkout_url  = $body['checkout_url'];
            $api_host      = parse_url($api_base, PHP_URL_HOST);
            $checkout_host = parse_url($checkout_url, PHP_URL_HOST);
            if ($api_host && $checkout_host !== $api_host) {
                $json['error'] = $this->language->get('error_payment_failed');
                $this->response->addHeader('Content-Type: application/json');
                $this->response->setOutput(json_encode($json));
                return;
            }
            // Stash token in session so callback() can verify payment on return
            preg_match('#/checkout/([A-Za-z0-9_-]+)$#', $checkout_url, $tm);
            $this->session->data['algovoi_token']    = isset($tm[1]) ? $tm[1] : ($body['id'] ?? '');
            $this->session->data['algovoi_order_id'] = $this->session->data['order_id'];
            $pending_status = (int)$this->config->get('payment_algovoi_pending_status_id') ?: 1;
            $this->model_checkout_order->addHistory($this->session->data['order_id'], $pending_status, 'Awaiting AlgoVoi payment', false);
            $json['redirect'] = $checkout_url;
        } else {
            $json['error'] = $this->language->get('error_payment_failed');
        }

        $this->response->addHeader('Content-Type: application/json');
        $this->response->setOutput(json_encode($json));
    }

    public function callback(): void {
        $lang     = $this->config->get('config_language');
        $token    = $this->session->data['algovoi_token'] ?? '';
        $order_id = $this->session->data['algovoi_order_id'] ?? ($this->session->data['order_id'] ?? 0);
        $api_base = rtrim($this->config->get('payment_algovoi_api_base_url') ?: 'https://cloud.algovoi.co.uk', '/');
        $api_key  = $this->config->get('payment_algovoi_admin_api_key');

        $paid = false;
        $tx_id = '';

        if ($token && $order_id && $api_base && strpos($api_base, 'https://') === 0) {
            $ch = curl_init($api_base . '/checkout/' . rawurlencode($token) . '/status');
            curl_setopt_array($ch, [
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_TIMEOUT        => 15,
                CURLOPT_SSL_VERIFYPEER => true,
                CURLOPT_SSL_VERIFYHOST => 2,
                CURLOPT_HTTPHEADER     => ['Authorization: Bearer ' . $api_key],
            ]);
            $resp      = curl_exec($ch);
            $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            curl_close($ch);

            if ($http_code === 200) {
                $data  = json_decode($resp, true);
                $status = $data['status'] ?? '';
                if (in_array($status, ['paid', 'completed', 'confirmed'], true)) {
                    $paid  = true;
                    $tx_id = $data['tx_id'] ?? '';
                }
            }
        }

        if ($paid) {
            $this->load->model('checkout/order');
            $order = $this->model_checkout_order->getOrder((int)$order_id);
            if ($order && in_array($order['order_status_id'], [1, 2])) {
                $complete_status = (int)$this->config->get('payment_algovoi_complete_status_id') ?: 5;
                $this->model_checkout_order->addHistory(
                    (int)$order_id,
                    $complete_status,
                    'AlgoVoi payment confirmed on return. TX: ' . $tx_id,
                    true
                );
            }
            unset($this->session->data['algovoi_token'], $this->session->data['algovoi_order_id']);
            $this->response->redirect($this->url->link('checkout/success', 'language=' . $lang, true));
        } else {
            // Payment was cancelled or not yet confirmed — send back to checkout with a notice
            unset($this->session->data['algovoi_token'], $this->session->data['algovoi_order_id']);
            $this->session->data['error'] = 'Payment was cancelled or not completed. Please try again.';
            $this->response->redirect($this->url->link('checkout/checkout', 'language=' . $lang, true));
        }
    }

    public function webhook(): void {
        $raw_body = file_get_contents('php://input');
        $secret   = $this->config->get('payment_algovoi_webhook_secret');

        // FIX: reject immediately if webhook secret is not configured
        if (empty($secret)) {
            http_response_code(500);
            exit('Webhook secret not configured');
        }

        $sig_header = $_SERVER['HTTP_X_ALGOVOI_SIGNATURE'] ?? '';
        $expected   = base64_encode(hash_hmac('sha256', $raw_body, $secret, true));
        if (!hash_equals($expected, $sig_header)) {
            http_response_code(401);
            exit('Unauthorized');
        }

        $data = json_decode($raw_body, true);
        if (empty($data['order_id'])) {
            http_response_code(400);
            exit('Bad Request');
        }

        $this->load->model('checkout/order');
        $order = $this->model_checkout_order->getOrder((int)$data['order_id']);

        if ($order && in_array($order['order_status_id'], [1, 2])) {
            $complete_status = (int)$this->config->get('payment_algovoi_complete_status_id') ?: 5;
            $this->model_checkout_order->addHistory(
                (int)$data['order_id'],
                $complete_status,
                'AlgoVoi payment confirmed. TX: ' . ($data['tx_id'] ?? 'n/a'),
                true
            );
        }

        http_response_code(200);
        echo json_encode(['status' => 'ok']);
        exit;
    }
}

File: native-go/algovoi.go
// Package algovoi provides a zero-dependency Go client for the AlgoVoi payment platform.
//
// Supports:
//   - Hosted checkout (Algorand, VOI, Hedera) — redirect to AlgoVoi payment page
//   - Extension payment (Algorand, VOI) — in-page wallet flow via algosdk
//   - Webhook verification with HMAC
//   - SSRF protection on checkout URL fetches
//   - Cancel-bypass prevention on hosted return
//
// No third-party dependencies — uses only the Go standard library.
//
// Version: 1.0.0
package algovoi

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"html"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

// AlgodConfig holds node configuration for a specific chain.
type AlgodConfig struct {
	URL     string
	AssetID int
	Ticker  string
	Dec     int
}

var algodConfigs = map[string]AlgodConfig{
	"algorand-mainnet": {URL: "https://mainnet-api.algonode.cloud", AssetID: 31566704, Ticker: "USDC", Dec: 6},
	"voi-mainnet":      {URL: "https://mainnet-api.voi.nodely.io", AssetID: 302190, Ticker: "aUSDC", Dec: 6},
}

var hostedNetworks = map[string]bool{
	"algorand_mainnet": true,
	"voi_mainnet":      true,
	"hedera_mainnet":   true,
	"stellar_mainnet":  true,
}
var extNetworks = map[string]bool{"algorand_mainnet": true, "voi_mainnet": true}

var tokenRegex = regexp.MustCompile(`/checkout/([A-Za-z0-9_-]+)$`)
var addrRegex = regexp.MustCompile(`<div[^>]+id=["']addr["'][^>]*>([A-Z2-7]{58})<`)
var memoRegex = regexp.MustCompile(`<div[^>]+id=["']memo["'][^>]*>(algovoi:[^<]+)<`)

// Client is the AlgoVoi payment adapter.
type Client struct {
	APIBase       string
	APIKey        string
	TenantID      string
	WebhookSecret string
	HTTPClient    *http.Client
}

// New creates a new AlgoVoi client with sensible defaults.
func New(apiBase, apiKey, tenantID, webhookSecret string) *Client {
	return &Client{
		APIBase:       strings.TrimRight(apiBase, "/"),
		APIKey:        apiKey,
		TenantID:      tenantID,
		WebhookSecret: webhookSecret,
		HTTPClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// ── Payment Link ────────────────────────────────────────────────────────

// PaymentLinkResponse is the API response from creating a payment link.
type PaymentLinkResponse struct {
	CheckoutURL      string `json:"checkout_url"`
	ID               string `json:"id"`
	Chain            string `json:"chain"`
	AmountMicrounits int    `json:"amount_microunits"`
	AssetID          int    `json:"asset_id"`
}

// CreatePaymentLink creates a payment link via the AlgoVoi API.
func (c *Client) CreatePaymentLink(amount float64, currency, label, network, redirectURL string) (*PaymentLinkResponse, error) {
	payload := map[string]interface{}{
		"amount":            amount,
		"currency":          strings.ToUpper(currency),
		"label":             label,
		"preferred_network": network,
	}
	if redirectURL != "" {
		payload["redirect_url"] = redirectURL
		payload["expires_in_seconds"] = 3600
	}

	body, err := c.post(c.APIBase+"/v1/payment-links", payload, true)
	if err != nil {
		return nil, err
	}

	var resp PaymentLinkResponse
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, fmt.Errorf("algovoi: invalid response: %w", err)
	}
	if resp.CheckoutURL == "" {
		return nil, fmt.Errorf("algovoi: empty checkout_url in response")
	}
	return &resp, nil
}

// ExtractToken extracts the short token from a checkout URL.
func ExtractToken(checkoutURL string) string {
	m := tokenRegex.FindStringSubmatch(checkoutURL)
	if len(m) < 2 {
		return ""
	}
	return m[1]
}

// ── Hosted Checkout ─────────────────────────────────────────────────────

// HostedResult contains the data from a hosted checkout initiation.
type HostedResult struct {
	CheckoutURL      string
	Token            string
	Chain            string
	AmountMicrounits int
}

// HostedCheckout starts a hosted checkout. Returns the redirect URL and token.
func (c *Client) HostedCheckout(amount float64, currency, label, network, redirectURL string) (*HostedResult, error) {
	if !hostedNetworks[network] {
		network = "algorand_mainnet"
	}

	link, err := c.CreatePaymentLink(amount, currency, label, network, redirectURL)
	if err != nil {
		return nil, err
	}

	return &HostedResult{
		CheckoutURL:      link.CheckoutURL,
		Token:            ExtractToken(link.CheckoutURL),
		Chain:            link.Chain,
		AmountMicrounits: link.AmountMicrounits,
	}, nil
}

// VerifyHostedReturn checks whether a hosted checkout was actually paid.
// Call this when the customer returns from the hosted checkout page.
//
// CRITICAL: Without this check, a customer can cancel payment and still
// appear to have paid (cancel-bypass vulnerability).
func (c *Client) VerifyHostedReturn(token string) (bool, error) {
	if token == "" {
		return false, nil
	}

	req, err := http.NewRequest("GET", c.APIBase+"/checkout/"+url.PathEscape(token), nil)
	if err != nil {
		return false, err
	}

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return false, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return false, nil
	}

	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return false, err
	}

	var data struct {
		Status string `json:"status"`
	}
	if err := json.Unmarshal(body, &data); err != nil {
		return false, nil
	}

	switch data.Status {
	case "paid", "completed", "confirmed":
		return true, nil
	default:
		return false, nil
	}
}

// ── Extension Payment ───────────────────────────────────────────────────

// ExtensionData contains all variables needed to render the JS payment UI.
type ExtensionData struct {
	Token         string
	Receiver      string
	Memo          string
	AmountMU      int
	AssetID       int
	AlgodURL      string
	Ticker        string
	AmountDisplay string
	Chain         string
	CheckoutURL   string
}

// ExtensionCheckout prepares data for the in-page wallet payment flow.
func (c *Client) ExtensionCheckout(amount float64, currency, label, network string) (*ExtensionData, error) {
	if !extNetworks[network] {
		network = "algorand_mainnet"
	}

	link, err := c.CreatePaymentLink(amount, currency, label, network, "")
	if err != nil {
		return nil, err
	}

	chain := link.Chain
	if chain == "" {
		chain = "algorand-mainnet"
	}
	amountMU := link.AmountMicrounits
	algod, ok := algodConfigs[chain]
	if !ok {
		algod = algodConfigs["algorand-mainnet"]
	}

	scraped, err := c.scrapeCheckout(link.CheckoutURL)
	if err != nil {
		return nil, err
	}

	token := ExtractToken(link.CheckoutURL)
	divisor := 1.0
	for i := 0; i < algod.Dec; i++ {
		divisor *= 10
	}

	return &ExtensionData{
		Token:         token,
		Receiver:      scraped.Receiver,
		Memo:          scraped.Memo,
		AmountMU:      amountMU,
		AssetID:       algod.AssetID,
		AlgodURL:      algod.URL,
		Ticker:        algod.Ticker,
		AmountDisplay: fmt.Sprintf("%.2f", float64(amountMU)/divisor),
		Chain:         chain,
		CheckoutURL:   link.CheckoutURL,
	}, nil
}

// VerifyExtensionPayment verifies an extension payment transaction with the AlgoVoi API.
func (c *Client) VerifyExtensionPayment(token, txID string) (map[string]interface{}, error) {
	if token == "" || txID == "" || len(txID) > 200 {
		return map[string]interface{}{"error": "Invalid parameters"}, fmt.Errorf("invalid parameters")
	}

	verifyURL := c.APIBase + "/checkout/" + url.PathEscape(token) + "/verify"
	body, err := c.postRaw(verifyURL, map[string]interface{}{"tx_id": txID})
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}
	return result, nil
}

// ── Webhook ─────────────────────────────────────────────────────────────

// VerifyWebhook verifies and parses an incoming webhook request.
// Returns nil if the secret is empty or the signature is invalid.
func (c *Client) VerifyWebhook(rawBody []byte, signature string) (map[string]interface{}, error) {
	if c.WebhookSecret == "" {
		return nil, fmt.Errorf("webhook secret not configured")
	}

	mac := hmac.New(sha256.New, []byte(c.WebhookSecret))
	mac.Write(rawBody)
	expected := base64.StdEncoding.EncodeToString(mac.Sum(nil))

	if !hmac.Equal([]byte(expected), []byte(signature)) {
		return nil, fmt.Errorf("invalid signature")
	}

	var payload map[string]interface{}
	if err := json.Unmarshal(rawBody, &payload); err != nil {
		return nil, fmt.Errorf("invalid JSON: %w", err)
	}
	return payload, nil
}

// ── HTTP Handler Helpers ────────────────────────────────────────────────

// WebhookHandler returns an http.HandlerFunc that verifies and processes webhooks.
// The callback receives the parsed payload on success.
func (c *Client) WebhookHandler(callback func(payload map[string]interface{})) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
			return
		}

		body, err := io.ReadAll(io.LimitReader(r.Body, 1<<20))
		if err != nil {
			http.Error(w, "Bad request", http.StatusBadRequest)
			return
		}

		signature := r.Header.Get("X-AlgoVoi-Signature")
		payload, err := c.VerifyWebhook(body, signature)
		if err != nil {
			http.Error(w, "Unauthorized", http.StatusUnauthorized)
			return
		}

		callback(payload)
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(`{"ok":true}`))
	}
}

// ── HTML Helpers ────────────────────────────────────────────────────────

// ChainInfo represents a chain option for the selector.
type ChainInfo struct {
	Value  string
	Label  string
	Ticker string
	Colour string
}

// RenderChainSelector renders chain selector radio buttons as HTML.
// mode should be "hosted" (4 chains: Algorand, VOI, Hedera, Stellar) or
// "extension" (2 chains: Algorand, VOI — the AlgoVoi browser extension signs
// those two; Hedera/Stellar buyers use hosted checkout with their own wallet).
func RenderChainSelector(fieldName, mode string) string {
	chains := []ChainInfo{
		{Value: "algorand_mainnet", Label: "Algorand", Ticker: "USDC", Colour: "#3b82f6"},
		{Value: "voi_mainnet", Label: "VOI", Ticker: "aUSDC", Colour: "#8b5cf6"},
	}
	if mode == "hosted" {
		chains = append(chains, ChainInfo{Value: "hedera_mainnet", Label: "Hedera", Ticker: "USDC", Colour: "#00a9a5"})
		chains = append(chains, ChainInfo{Value: "stellar_mainnet", Label: "Stellar", Ticker: "USDC", Colour: "#7C63D0"})
	}

	var sb strings.Builder
	sb.WriteString(`<div style="margin:.5rem 0;font-size:12px;color:#6b7280;font-weight:600;text-transform:uppercase;letter-spacing:.04em;">Select network</div>`)
	sb.WriteString(`<div style="display:flex;gap:16px;flex-wrap:wrap;margin-bottom:.5rem;">`)
	for i, c := range chains {
		checked := ""
		if i == 0 {
			checked = " checked"
		}
		sb.WriteString(fmt.Sprintf(
			`<label style="display:flex;align-items:center;gap:6px;cursor:pointer;font-size:13px;">`+
				`<input type="radio" name="%s" value="%s"%s style="accent-color:%s;">`+
				` %s &mdash; %s</label>`,
			html.EscapeString(fieldName), html.EscapeString(c.Value), checked, c.Colour,
			html.EscapeString(c.Label), html.EscapeString(c.Ticker),
		))
	}
	sb.WriteString(`</div>`)
	return sb.String()
}

// IsValidHostedNetwork returns true if the network is valid for hosted checkout.
func IsValidHostedNetwork(network string) bool { return hostedNetworks[network] }

// IsValidExtNetwork returns true if the network is valid for extension payment.
func IsValidExtNetwork(network string) bool { return extNetworks[network] }

// ── Internal ────────────────────────────────────────────────────────────

func (c *Client) post(url string, data map[string]interface{}, auth bool) ([]byte, error) {
	payload, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("POST", url, bytes.NewReader(payload))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if auth {
		req.Header.Set("Authorization", "Bearer "+c.APIKey)
		req.Header.Set("X-Tenant-Id", c.TenantID)
	}

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return nil, err
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("algovoi: HTTP %d: %s", resp.StatusCode, string(body))
	}
	return body, nil
}

func (c *Client) postRaw(rawURL string, data map[string]interface{}) ([]byte, error) {
	payload, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("POST", rawURL, bytes.NewReader(payload))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	return io.ReadAll(io.LimitReader(resp.Body, 1<<20))
}

type scrapeResult struct {
	Receiver string
	Memo     string
}

func (c *Client) scrapeCheckout(checkoutURL string) (*scrapeResult, error) {
	// SSRF guard: host must match API base
	apiHost := mustParseHost(c.APIBase)
	checkoutHost := mustParseHost(checkoutURL)
	if apiHost == "" || checkoutHost != apiHost {
		return nil, fmt.Errorf("algovoi: checkout URL host mismatch (SSRF blocked)")
	}

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Get(checkoutURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return nil, err
	}
	htmlStr := string(body)

	var receiver, memo string
	if m := addrRegex.FindStringSubmatch(htmlStr); len(m) >= 2 {
		receiver = m[1]
	}
	if m := memoRegex.FindStringSubmatch(htmlStr); len(m) >= 2 {
		memo = strings.TrimSpace(m[1])
	}

	if receiver == "" || memo == "" {
		return nil, fmt.Errorf("algovoi: could not extract receiver/memo from checkout page")
	}
	return &scrapeResult{Receiver: receiver, Memo: memo}, nil
}

func mustParseHost(rawURL string) string {
	u, err := url.Parse(rawURL)
	if err != nil {
		return ""
	}
	return u.Hostname()
}
